core/src/main/scala/org/apache/spark/HeartbeatReceiver.scala

```diff
@@ -77,17 +77,61 @@ private[spark] class HeartbeatReceiver(sc: SparkContext, clock: Clock)
 
   private[spark] var scheduler: TaskScheduler = null
 
-  // executor ID -> timestamp of when the last heartbeat from this executor was received
+  /**
+   * [SPARK-39984]
+   * Please make sure the intersection between `executorLastSeen` and `executorExpiryCandidates` is
+   * an empty set. If the intersection is not empty, it is possible to never kill the executor until
+   * the executor recovers. When an executor is in both `executorLastSeen` and
+   * `executorExpiryCandidates`, the value of `workerLastHeartbeat` in `executorExpiryCandidates`
+   * may update if the worker sends heartbeats to master normally.
+   *
+   * `executorLastSeen`:
+   *  - key: executor ID
+   *  - value: timestamp of when the last heartbeat from this executor was received
+   *
+   * `executorExpiryCandidates`: executor ID -> WorkerLastHeartbeat
+   *  - key: executor ID
+   *  - value: timestamp of when the last heartbeat from the worker was received
+   *
+   * when driver does not receive any heartbeat from an executor for `executorTimeoutMs` seconds,
+   * the driver will ask master for the last heartbeat from the worker which the executor is running
+   * on.
+   */
   private val executorLastSeen = new HashMap[String, Long]
+  private val executorExpiryCandidates = new HashMap[String, Long]
 
   private val executorTimeoutMs = sc.conf.get(
     config.STORAGE_BLOCKMANAGER_HEARTBEAT_TIMEOUT
-  ).getOrElse(Utils.timeStringAsMs(s"${sc.conf.get(Network.NETWORK_TIMEOUT)}s"))
+  ).getOrElse(
+    sc.conf.get(Network.NETWORK_EXECUTOR_TIMEOUT) match {
+      case Some(executorTimeout) => executorTimeout
+      case None => Utils.timeStringAsMs(s"${sc.conf.get(Network.NETWORK_TIMEOUT)}s")
+    }
+  )
 
   private val checkTimeoutIntervalMs = sc.conf.get(Network.NETWORK_TIMEOUT_INTERVAL)
 
   private val executorHeartbeatIntervalMs = sc.conf.get(config.EXECUTOR_HEARTBEAT_INTERVAL)
 
+  /**
+   * Currently, [SPARK-39984] is only for StandaloneSchedulerBackend.
+   *
+   * `checkWorkerLastHeartbeat`: A flag to enable two-phase executor timeout.
+   * `expiryCandidatesTimeout`: The timeout used for executorExpiryCandidates.
+   */
+  private val checkWorkerLastHeartbeat = {
```

Comment:
Shall we make this a lazy val? Then, we can check `sc.schedulerBackend.isInstanceOf[StandaloneSchedulerBackend]` directly here instead of calling a fucntion.

Reply:
This is a good point, but there is a question. 

In my understanding, your comment is as following code. 
```scala
  private lazy val checkWorkerLastHeartbeat =
    sc.conf.get(config.HEARTBEAT_RECEIVER_CHECK_WORKER_LAST_HEARTBEAT)
      && sc.schedulerBackend.isInstanceOf[StandaloneSchedulerBackend]
```
In the function `expireDeadHosts`, we will check the value of `checkWorkerLastHeartbeat`, so this variable will be evaluated. However, if the first `expireDeadHosts` trigger is prior to scheduler backend initialization, the value of `checkWorkerLastHeartbeat` will be false (`sc.schedulerBackend.isInstanceOf[StandaloneSchedulerBackend]` == false). If the scheduler backend initialization finishes before the second `expireDeadHosts` trigger, the value of `checkWorkerLastHeartbeat` should be true. Nevertheless, it is set to false in the first `expireDeadHosts` function call.


Maybe the other solution is to define `checkWorkerLastHeartbeat` as a function rather than a variable. Any thoughts?

Reply:
> However, if the first expireDeadHosts trigger is prior to scheduler backend initialization, the value of checkWorkerLastHeartbeat will be false

I thought there would be an initial delay for the first `expireDeadHosts` so `checkWorkerLastHeartbeat` is less likely to return false. But seems it doesn't have the initial delay, though I think it doesn't make sense.

I actually think we can set the initial delay to its check interval: `checkTimeoutIntervalMs`.

Reply:
Let me clarify the definition of "initial delay". In my understanding, the "initial delay" above means the second argument of `scheduleAtFixedRate` ([Link](https://sourcegraph.com/github.com/apache/spark@b012cb7/-/blob/core/src/main/scala/org/apache/spark/HeartbeatReceiver.scala?L107)). Is it correct?

```scala
override def onStart(): Unit = {
  timeoutCheckingTask = eventLoopThread.scheduleAtFixedRate(
  () => Utils.tryLogNonFatalError { Option(self).foreach(_.ask[Boolean](ExpireDeadHosts)) },
  0, checkTimeoutIntervalMs, TimeUnit.MILLISECONDS)
}
```

I agreed with you that initial delay can significantly decrease the probability of first expireDeadHosts trigger happening before scheduler backend initialization. However, it cannot totally prevent this edge case. Hence, my thought is to define
`checkWorkerLastHeartbeat` as a function, and we can totally prevent the edge case. Does it make sense?

Reply:
> However, it cannot totally prevent this edge case

Theoretically, I agree it doesn't but given that checkTimeoutIntervalMs by default is 60s it's almost impossible to have scheduler backend not initialized when the first expireDeadHosts raised.


> Hence, my thought is to define
checkWorkerLastHeartbeat as a function, and we can totally prevent the edge case. Does it make sense?

I don't think it makes a big difference with the already defined `isStandalone()` function. Maybe just leave it as is if we have no better idea.

Reply:
Updated. 

(1) Declare the variable `checkWorkerLastHeartbeat` as a lazy variable
(2) Set an initial delay for `ExpireDeadHosts`
(3) If the scheduler backend has not initialized when `checkWorkerLastHeartbeat` is evaluated, an `UnsupportedOperationException` will be thrown.
```scala
private lazy val checkWorkerLastHeartbeat = sc.schedulerBackend match {
  case _: CoarseGrainedSchedulerBackend =>
    sc.conf.get(config.HEARTBEAT_RECEIVER_CHECK_WORKER_LAST_HEARTBEAT) &&
      sc.schedulerBackend.isInstanceOf[StandaloneSchedulerBackend]
  case _: LocalSchedulerBackend => false
  case other => throw new UnsupportedOperationException(
    s"Unknown scheduler backend: ${other.getClass}")
}
```